### PR TITLE
[X86] Support -march=diamondrapids

### PR DESCRIPTION
**Pull request by FreddyLeaf**
Ref.: https://cdrdv2.intel.com/v1/dl/getContent/671368

### CODE REVIEW

File: clang/test/Preprocessor/predefined-arch-macros.c [hunk 1]
```diff
@@ -1867,13 +1867,18 @@
 // RUN: %clang -march=graniterapids-d -m32 -E -dM %s -o - 2>&1 \
 // RUN:     --target=i386 \
 // RUN:   | FileCheck -match-full-lines %s -check-prefixes=CHECK_GNR_M32,CHECK_GNRD_M32
+// RUN: %clang -march=diamondrapids -m32 -E -dM %s -o - 2>&1 \
+// RUN:     --target=i386 \
+// RUN:   | FileCheck -match-full-lines %s -check-prefixes=CHECK_GNR_M32,CHECK_GNRD_M32,CHECK_DMR_M32
```
Comment (RKSimon) [inline clang/test/Preprocessor/predefined-arch-macros.c:1872]: This 'overlay' approach feels a bit unwieldy - but I have not strong objections.

File: compiler-rt/lib/builtins/cpu_model/x86.c [hunk 2]
```diff
@@ -600,6 +601,19 @@ static const char *getIntelProcessorTypeAndSubtype(unsigned Family,
...
+  case 19:
+    switch (Model) {
+    // Diamondrapids:
+    case 0x01:
+      CPU = "diamondrapids";
+      *Type = INTEL_COREI7;
+      *Subtype = INTEL_COREI7_DIAMONDRAPIDS;
+      break;
+
+    default: // Unknown family 0x13 CPU.
```
Comment (RKSimon) [inline compiler-rt/lib/builtins/cpu_model/x86.c:613]: Inconsistent use of fam 19 and 0x13?

File: llvm/lib/Target/X86/X86.td [hunk 3]
```diff
@@ -1139,6 +1139,28 @@ def ProcessorFeatures {
   list<SubtargetFeature> GNRDFeatures =
     !listconcat(GNRFeatures, GNRDAdditionalFeatures);
 
+  // Diamondrapids
```
Comment (RKSimon) [inline llvm/lib/Target/X86/X86.td:1142]: (trivial) Diamondrapids -> DiamondRapids (for consistency)

**Review by RKSimon — COMMENTED** [review]
Mechanically this LGTM with a few style minors, but I don't know much about the exact arch/ISA so please wait for @phoebewang's review

File: llvm/lib/Target/X86/X86.td (hunk 3 above)
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86.td]: [52e63a0](https://github.com/llvm/llvm-project/pull/113881/commits/52e63a0f4b48ad106471343f2f48976bd9b84661)

File: compiler-rt/lib/builtins/cpu_model/x86.c (hunk 2 above)
Comment (FreddyLeaf) [reply compiler-rt/lib/builtins/cpu_model/x86.c]: [52e63a0](https://github.com/llvm/llvm-project/pull/113881/commits/52e63a0f4b48ad106471343f2f48976bd9b84661)

File: llvm/include/llvm/TargetParser/X86TargetParser.h [hunk 4]
```diff
@@ -121,6 +121,7 @@ enum CPUKind {
   CK_GraniterapidsD,
   CK_Emeraldrapids,
   CK_Clearwaterforest,
+  CK_DiamondRapids,
```
Comment (phoebewang) [inline llvm/include/llvm/TargetParser/X86TargetParser.h:124]: CK_Diamondrapids?

File: llvm/lib/TargetParser/X86TargetParser.cpp [hunk 5]
```diff
@@ -381,6 +389,8 @@ constexpr ProcInfo Processors[] = {
   { {"emeraldrapids"}, CK_Emeraldrapids, FEATURE_AVX512FP16, FeaturesSapphireRapids, 'n', false },
   // Clearwaterforest microarchitecture based processors.
   { {"clearwaterforest"}, CK_Lunarlake, FEATURE_AVX2, FeaturesClearwaterforest, 'p', false },
+  // Diamond Rapids microarchitecture based processors.
+  { {"diamondrapids"}, CK_DiamondRapids, FEATURE_AVX10_2_512, FeaturesDiamondRapids, 'z', false },
```
Comment (phoebewang) [inline llvm/lib/TargetParser/X86TargetParser.cpp:393]: ditto.

File: llvm/lib/Target/X86/X86.td [hunk 6]
```diff
@@ -1155,6 +1155,34 @@ def ProcessorFeatures {
   list<SubtargetFeature> GNRDFeatures =
     !listconcat(GNRFeatures, GNRDAdditionalFeatures);
 
+  // Diamond Rapids
+  list<SubtargetFeature> DMRAdditionalFeatures = [FeatureAVX10_2_512,
+                                                  FeatureAMXCOMPLEX,
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86.td:1160]: `FeatureAMXCOMPLEX` alreay inherited from `GNRDFeatures`.

File: llvm/include/llvm/TargetParser/X86TargetParser.h (hunk 4 above)
Comment (FreddyLeaf) [reply llvm/include/llvm/TargetParser/X86TargetParser.h:124]: I prefer your style, too. Let's refine CK_SapphireRapids as well in the future

File: llvm/lib/Target/X86/X86.td (hunk 6 above)
Comment (FreddyLeaf) [reply llvm/lib/Target/X86/X86.td]: ab477fb

File: llvm/lib/TargetParser/X86TargetParser.cpp (hunk 5 above)
Comment (FreddyLeaf) [reply llvm/lib/TargetParser/X86TargetParser.cpp]: ab477fb

File: llvm/include/llvm/TargetParser/X86TargetParser.h (hunk 4 above)
Comment (FreddyLeaf) [reply llvm/include/llvm/TargetParser/X86TargetParser.h]: ab477fb

**Review by phoebewang — APPROVED** [review]
LGTM.

**Comment by nico** [comment]
Looks like this breaks tests: http://45.33.8.238/linux/153081/step_6.txt

Please take a look and revert for now if it takes a while to fix.

**Comment by FreddyLeaf** [comment]
> Looks like this breaks tests: http://45.33.8.238/linux/153081/step_6.txt
> 
> Please take a look and revert for now if it takes a while to fix.

Thanks reminding. Revert first.

**Pull request merged**